### PR TITLE
Add `TransactionManager` to execute certified transactions based on input object availability

### PR DESCRIPTION
**Pull request by mwtian**
### Background

This change adds a `TransactionManager` component that manages dependencies between pending certified transactions and objects in their input. `TransactionManager` accepts certified transactions from various sources e.g. consensus output currently and checkpoint sync in future. It builds dependencies between objects and transactions, and publishes transactions that have all input available locally. The goal of the change is to avoid the inefficiency and instability in the existing execution logic, by guaranteeing availability of input objects before execution, and simplify execution logic.                      

### Transaction Manager

Here is a diagram for how `TransactionManager` communicates with other components.

transaction_manager drawio

Right now, `TransactionManager` is only activated in validators. The next step is to use `TransactionManager` in fullnode as well.

### Shared locks

Since `TransactionManager` requires input objects' versions, shared objects in a certificate must be locked before sending to `TransactionManager`. Currently, `SharedObjectLockNotSetError` is often observed in logs, so this is not yet guaranteed. Enforcing this invariant involves a few changes:
1. Drop the logic to remove shared object locks after a transaction commits. I'm unable to convince myself the existing logic of removing shared locks are correct. And even with 2 and 3 below, I still observed `SharedObjectLockNotSetError` in tests. Later after dust settles down with the new `TransactionManager`, we can implement the logic to clear shared locks table at epoch boundaries.
2. In `handle_consensus_transaction()`, take shared locks and add certificates to the `pending_certificates` table in one batch write, then send the certificate to `TransactionManager`. Otherwise if a certificate is added to the pending queue or TransactionManager first, it can get executed immediately, without taking the shared locks.
3. Certificates were sent to the pending queue (`add_pending_certificates()`) quite liberally before. Most of these corner cases are removed to avoid sending a certificate containing shared objects to `TransactionManager`, if no shared locks are taken for the certificate.

### Execution Driver and Node Sync

The new execution driver runs `AuthorityState::handle_certificate()` to process certificates directly, instead of sending digests to node sync. The driver shares similarities with the execution logic in node sync. Eventually we may want to merge the two. But right now, keeping the validator execution logic simple seems to be more beneficial.

There is no longer a retry path for certificates missing parents, since it is no longer possible. We will rely on other mechanisms (e.g. Narwhal) to catch up on missing certificates.

#4990 #5514

### CODE REVIEW

File: crates/sui-core/src/transaction_manager.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+use crate::authority::{authority_store::ObjectKey, AuthorityMetrics, AuthorityStore};
+
+/// TransactionManager is responsible for managing pending certificates and publishes a stream
+/// of certificates ready to be executed. It works together with AuthorityState for receiving
+/// pending certificates, and getting notified about committed objects. Executing driver
+/// subscribes to the stream of ready certificates published by the TransactionManager, and can
+/// execute them in parallel.
+/// TODO: use TransactionManager for fullnode.
+pub(crate) struct TransactionManager {
```
Comment (mystenmark) [inline crates/sui-core/src/transaction_manager.rs:25]: A quick sketch of a design for handling unsequenced certs:

1. pending certs can be in one of three states: Unsequenced, InputsMissing, and Executable.
2. Certs start in the Unsequenced state.
3. If they have no shared objects, they advance to InputsMissing.
4. If they have shared objects, we call `is_certificate_awaiting_sequencing()` to determine whether they advance to InputMissing.
5. The consensus task notifies TransactionManager when a cert has been sequenced, so that it can advance to InputsMissing.
6. The InputsMissing->Executable transition is of course the piece that you've already implemented.

File: crates/sui-core/src/transaction_manager.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        };
+        transaction_manager
+            .enqueue(node_sync_store.all_pending_certs().unwrap())
+            .expect("Initialize TransactionManager with pending certificates failed.");
+        transaction_manager
+    }
+
+    /// Enqueues a certificate into TransactionManager. Once all of its input objects are available
+    /// locally, it will be published.
+    /// TODO: turn this function async, and wait for inflight certificates below threshold.
```
Comment (mystenmark) [inline crates/sui-core/src/transaction_manager.rs:59]: I don't think this TODO is possible - if we bound the number of inflight certificates, we may deadlock, because we may be awaiting the completion of a cert whose dependencies we are trying to enqueue.

File: crates/sui-core/src/transaction_manager.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            if set.is_empty() {
+                self.pending_certificates.remove(&cert_key);
+                // NOTE: failing and ignoring the certificate is fine, if it will be retried at a higher level.
+                // Otherwise, this has to crash.
+                let cert = match self.node_sync_store.get_cert(cert_key.0, &cert_key.1) {
+                    Ok(Some(cert)) => cert,
+                    Ok(None) => {
+                        error!(
+                            "Ready certificate not found in the pending table: {:?}",
+                            cert_key
```
Comment (mystenmark) [inline crates/sui-core/src/transaction_manager.rs:110]: Please log as `error!(tx_digest = ?cert_key, "Ready certificate not found in the pending table")`

File: crates/sui-core/src/transaction_manager.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                        );
+                        continue;
+                    }
+                    Err(e) => {
+                        error!(
+                            "Failed to read pending table: key={:?}, err={}",
+                            cert_key, e
```
Comment (mystenmark) [inline crates/sui-core/src/transaction_manager.rs:117]: Same here

File: crates/sui-core/src/unit_tests/consensus_adapter_tests.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Mysten Labs, Inc.
```
Comment (mystenmark) [inline crates/sui-core/src/unit_tests/consensus_adapter_tests.rs:1]: I don't understand why this file is part of this PR.

File: crates/sui-core/src/unit_tests/consensus_tests.rs [hunk 6]
```diff
@@ -118,7 +118,7 @@ async fn listen_to_sequenced_transaction() {
         .unwrap();
 
     // Spawn a consensus listener.
-    ConsensusListener::spawn(/* rx_consensus_input */ rx_sui_to_consensus);
+    ConsensusListener::spawn(/* rx_consensus_listener */ rx_sui_to_consensus);
```
Comment (mystenmark) [inline crates/sui-core/src/unit_tests/consensus_tests.rs:121]: I think the changes in this file should probably be reverted.

File: crates/sui-storage/src/node_sync_store.rs [hunk 7]
```diff
@@ -24,12 +24,14 @@ use tracing::trace;
...
-/// NodeSyncStore store is used by nodes to store downloaded objects (pending_certs, etc) that have
+/// NodeSyncStore store is used by nodes to store received certificates and effects that have
 /// not yet been applied to the node's SuiDataStore.
 #[derive(DBMapUtils)]
 pub struct NodeSyncStore {
-    /// Certificates that have been fetched from remote validators, but not sequenced.
+    /// Certificates that have been received from clients, received from consensus and fetched from
+    /// remote validators, but not yet sequenced and executed.
     /// Entries are cleared after execution.
+    /// TODO: remove this table eventually, and replay the certificates from upstream for recovery.
```
Comment (mystenmark) [inline crates/sui-storage/src/node_sync_store.rs:34]: What does "from upstream" mean?

File: crates/sui-core/src/authority.rs [hunk 8]
```diff
@@ -905,6 +953,14 @@ impl AuthorityState {
         // commit_certificate finished, the tx is fully committed to the store.
         tx_guard.commit_tx();
 
+        // Notifies transaction manager about available input objects. This allows the transaction
+        // manager to schedule ready transactions.
+        // TODO: switch TransactionManager to use notify_read() on objects table.
```
Comment (mystenmark) [inline crates/sui-core/src/authority.rs:958]: I'm not sure notify_read is really a better approach - we could potentially have a massive number of waiting readers.

In any case, this should be moved to before `tx_guard.commit_tx()` to ensure at-least-once execution.

File: crates/sui-core/src/authority_active/execution_driver/mod.rs [hunk 9]
```diff
@@ -1,46 +1,49 @@
...
+};
+use tracing::{debug, error, info, warn};
 
 #[cfg(test)]
 pub(crate) mod tests;
 
+// Maximum number of certificates to be executed concurrently.
+// TODO: make this configurable.
+const MAX_EXECUTION_CONCURRENCY: usize = 20;
+const MAX_EXECUTION_TIMEOUT: Duration = Duration::from_secs(30);
```
Comment (mystenmark) [inline crates/sui-core/src/authority_active/execution_driver/mod.rs:26]: I don't believe we need a timeout here. We certainly have never observed a case of certificate execution hanging, and if we did it would be a massive bug that we would need to understand and fix. In addition, there are very few true `await` points in the execution path (most of them are compiled out) so a timeout can't even help very much.

The only place I can think of where a long wait is somewhat plausible is when we call `acquire_tx_guard()` - I say `somewhat` because any holder of the tx lock should finish quickly, therefore we should never wait very long. But if we were truly worried about that, we should put the timeout around `acquire_tx_guard()`, not here.

File: crates/sui-core/src/authority_active/execution_driver/mod.rs [hunk 10]
```diff
@@ -1,46 +1,49 @@
...
+    time::{sleep, timeout},
+};
+use tracing::{debug, error, info, warn};
 
 #[cfg(test)]
 pub(crate) mod tests;
 
+// Maximum number of certificates to be executed concurrently.
+// TODO: make this configurable.
+const MAX_EXECUTION_CONCURRENCY: usize = 20;
```
Comment (mystenmark) [inline crates/sui-core/src/authority_active/execution_driver/mod.rs:25]: Let's use num_cpus (https://docs.rs/num_cpus/latest/num_cpus/) as the default instead of hard coding a value.

File: crates/sui-core/src/transaction_manager.rs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+/// pending certificates, and getting notified about committed objects. Executing driver
+/// subscribes to the stream of ready certificates published by the TransactionManager, and can
+/// execute them in parallel.
+/// TODO: use TransactionManager for fullnode.
+pub(crate) struct TransactionManager {
+    authority_store: Arc<AuthorityStore>,
+    node_sync_store: Arc<NodeSyncStore>,
+    missing_inputs: BTreeMap<ObjectKey, (EpochId, TransactionDigest)>,
+    pending_certificates: BTreeMap<(EpochId, TransactionDigest), BTreeSet<ObjectKey>>,
+    tx_ready_certificates: UnboundedSender<VerifiedCertificate>,
```
Comment (mystenmark) [inline crates/sui-core/src/transaction_manager.rs:30]: I'm a bit worried about having a hidden, unbounded queue here, but I suppose its ok for now, especially since you are monitoring its size.

Eventually we need backpressure to allow us to reject user certs when we are overloaded, but that can be done later.

File: crates/sui-core/src/authority/authority_store.rs [hunk 12]
```diff
@@ -338,17 +259,19 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
         }
     }
 
+    #[allow(clippy::type_complexity)]
     pub fn get_sequenced_input_objects(
         &self,
         digest: &TransactionDigest,
         objects: &[InputObjectKind],
-    ) -> Result<Vec<Object>, SuiError> {
+    ) -> Result<(Vec<Object>, Vec<ObjectKey>, Vec<SuiError>), SuiError> {
```
Comment (mystenmark) [inline crates/sui-core/src/authority/authority_store.rs:267]: We should add a new function to return the missing objects, so that we can avoid reading and deserializing objects from withing TransactionManager. Also, this function is already very hard to understand, and a frequent cause of confusion and bugs due to the errors that it returns, so I would strongly prefer not to add any new complication here (even if that means we have to duplicate a small amount of code).

**Review by mystenmark — CHANGES_REQUESTED** [review]
Overall, this is fantastic! To summarize the important points from my comments:

- We need to be deliberate about how pending certificates are persisted and when it is ok to forget about them. This code seems to muddy those waters slightly.
- I believe TransactionManager should be aware of the unsequenced state and model it explicitly.
- I think there are a few data consistency problems here, (deletion of shared locks, notification of object writes).

File: crates/sui-core/src/authority_active/execution_driver/mod.rs [hunk 13]
```diff
@@ -51,202 +54,54 @@ pub async fn execution_process<A>(active_authority: Arc<ActiveAuthority<A>>)
...
-                seen.insert(*digest);
-                true
+        let limit = limit.clone();
+        // hold semaphore permit until task completes. unwrap ok because we never close
+        // the semaphore in this context.
+        let permit = limit.acquire_owned().await.unwrap();
+        let authority = active_authority.clone();
+
+        spawn_monitored_task!(async move {
+            let _ = &permit;
```
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-core/src/authority_active/execution_driver/mod.rs]: Good point! Fixed.

File: crates/sui-core/src/authority_active/execution_driver/mod.rs (hunk 9 above)
Comment (mwtian) [reply crates/sui-core/src/authority_active/execution_driver/mod.rs]: I took the timeout behavior from node sync, but I agree probably we don't need it here. Handling a certificate that timed out is really strange too. Removed the timeout for now.

**Review by mwtian — COMMENTED** [review]
Addressed ~half of the comments. Working on the rest.

File: crates/sui-core/src/authority_active/execution_driver/mod.rs [hunk 14]
```diff
@@ -51,202 +54,54 @@ pub async fn execution_process<A>(active_authority: Arc<ActiveAuthority<A>>)
...
-                info!(
-                    ?seq,
-                    ?digest,
-                    "serial certificate execution failed: {}",
-                    err
-                );
-            }
-        }
+            // Remove the pending certificate regardless of execution status. It can be retried
+            // from a higher level.
```
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-core/src/authority_active/execution_driver/mod.rs:132]: Thinking about this more, added a retry for temporary errors, and assume there will not be permanent errors.

File: crates/sui-core/src/transaction_manager.rs [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+        self.metrics
+            .transaction_manager_num_missing_objects
+            .set(self.missing_inputs.len() as i64);
+        self.metrics
+            .transaction_manager_num_pending_certificates
+            .set(self.pending_certificates.len() as i64);
+    }
+
+    /// Currently we are not yet confident that TransactionManager will be notified about all missing
+    /// objects. So we will run a periodic scanning task that checks if any input object is in
```
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-core/src/transaction_manager.rs:136]: Can I think about this more and update the comment as a followup? Is the following situation possible?
1. Transaction X notified TransactionManager about an object A before committing, but TransactionManager does not have object A as input yet.
2. TransactionManager enqueues a transaction Y that has A as input. This is ok before the new transaction Y has a different retry log.
3. TransactionManager will never get notification about A in future.

File: crates/sui-core/src/transaction_manager.rs (hunk 1 above)
Comment (mwtian) [reply crates/sui-core/src/transaction_manager.rs:24]: This should be implementable, but I don't see the benefit yet and it does add complexity. Can we keep the TransactionManager simple for now, until we need to support more use cases e.g. for checkpoint sync output?

Comment (mwtian) [reply crates/sui-core/src/transaction_manager.rs:24]: Added a todo to investigate taking shared object locks inside TransactionManager.

**Comment by mwtian** [comment]
> Overall, this is fantastic! To summarize the important points from my comments:
> 
> * We need to be deliberate about how pending certificates are persisted and when it is ok to forget about them. This code seems to muddy those waters slightly.
> * I believe TransactionManager should be aware of the unsequenced state and model it explicitly.
> * I think there are a few data consistency problems here, (deletion of shared locks, notification of object writes).

Thanks for very detailed and informative review! There are indeed issues I missed at the beginning, e.g. `pending_execution` / `pending_certificates` table is actually part of the data consistency and recovery guarantee. I moved writing to the `pending_certificates` table into the same batch write that updates consensus tables and execution index.

For shared locks, I understand it should mostly work. But we are indeed seeing `SharedObjectLockNotSetError` from logs currently. And even after implementing (2) and (3) from the shared lock section above (still removing shared locks after commit), I still observed the error within `TransactionManager::enqueue()`. There are code paths that make thinking about the correctness of shared locks difficult. For example in `hanele_certificate_and_effects()` no transaction recovery lock is taken, so many instances can run in parallel. When effects of the transaction exist, no shared lock is taken regardless of existence of shared locks. It may take a few days to make everything work correctly all the day.

And I don't see the benefit of clearing shared locks right now. They should not occupy as much disk space as some other data. I think cleaning up temporary transaction data can be sequenced to be worked on, after we have a solid reconfiguration implementation. Then some of the storage cleanup work will be trivial as part of the epoch cleanup. Also, I think we need proper rocksdb transaction or our own custom transaction implementation. Otherwise it is really error prone to manage data consistency with multiple reads and writes.

Let me know what you think and we can chat off thread.

File: crates/sui-core/src/transaction_manager.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-core/src/transaction_manager.rs]: As chatted yesterday, making sure notifications to TransactionManager happen after writes to object table are committed, and having the Mutex around access to TransactionManager, could be equivalent to notifiy_read() on the objects table. Updated comment.

File: crates/sui-core/src/authority.rs [hunk 16]
```diff
@@ -1590,36 +1682,13 @@ impl AuthorityState {
...
-    }
-
-    fn add_pending_impl(
-        &self,
-        certs: Vec<(TransactionDigest, Option<VerifiedCertificate>)>,
-        is_sequenced: bool,
-    ) -> SuiResult {
+    /// Adds certificates to the pending certificate store and transaction manager for ordered execution.
+    pub async fn add_pending_certificates(&self, certs: Vec<VerifiedCertificate>) -> SuiResult<()> {
+        // TODO: skip storing the certs and remove pending_certs table eventually.
```
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-core/src/authority.rs]: As discussed yesterday, relying on consensus output and checkpoint sync to replay pending transactions are possible, but add complexities. For now, we will just rely on using a pending_certificates table to hold certificates after they come out of other sources.

**Comment by mwtian** [comment]
@mystenmark as we discussed yesterday, I looked into how to make sure shared object locks can be written and cleared correctly per transaction, instead of per epoch as I proposed in this PR. I believe we must cover writing and clearing the shared object locks within a per-transaction critical section. I can think of two ways to implement this:
1. Reuse the transaction lock mechanism we have. Since the lock is a scoped object right now, we need to acquire and pass it through TransactionManager. And we may need to avoid waiting on the lock on critical path, e.g. `handle_consensus_transaction()`.
2. Simplify transaction locking by try inserting pending certificate. Transaction recovery will just be retrying pending certificates as well. This should be correct since we need to clear the pending certificate for each certificate, similar to unlocking the transaction. And this would simplify the transaction logic further.

What do you think @mystenmark @lxfind?

File: crates/sui-core/src/authority.rs [hunk 17]
```diff
@@ -1432,6 +1487,7 @@ impl AuthorityState {
 
     // TODO: This function takes both committee and genesis as parameter.
     // Technically genesis already contains committee information. Could consider merging them.
+    #[allow(clippy::disallowed_methods)]
```
Comment (mystenmark) [inline crates/sui-core/src/authority.rs:1490]: why was this necessary?

File: crates/sui-core/src/authority/authority_store.rs [hunk 18]
```diff
@@ -343,7 +332,59 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
...
+                                missing.push(ObjectKey(*id, *version));
+                            }
+                        }
+                        None => {
+                            // Abort the function because the lock should have been set.
+                            return Err(SuiError::SharedObjectLockNotSetError);
+                        }
+                    };
+                }
+                InputObjectKind::MovePackage(id) => {
```
Comment (mystenmark) [inline crates/sui-core/src/authority/authority_store.rs:367]: assuming that this function only runs for certs, this should be impossible - the cert cannot have been formed if immutable inputs were missing. this may be worth a comment.

File: crates/sui-core/src/authority/authority_store.rs [hunk 19]
```diff
@@ -1211,29 +1119,10 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
...
-        if !self.effects_exists(&transaction_digest)? {
-            write_batch = write_batch.insert_batch(
-                &self.epoch_tables().assigned_object_versions,
-                sequenced_to_write,
-            )?;
-        }
+        write_batch = write_batch.insert_batch(
+            &self.epoch_tables().assigned_object_versions,
+            sequenced_to_write,
+        )?;
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/authority/authority_store.rs:1242]: ok but let's make sure we don't forget

File: crates/sui-core/src/transaction_manager.rs [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+            .enqueue(node_sync_store.all_pending_certs().unwrap())
+            .expect("Initialize TransactionManager with pending certificates failed.");
+        transaction_manager
+    }
+
+    /// Enqueues a certificate into TransactionManager. Once all of its input objects are available
+    /// locally, it will be published.
+    /// TODO: turn this function async, and wait for inflight certificates below threshold.
+    pub(crate) fn enqueue(&mut self, certs: Vec<VerifiedCertificate>) -> SuiResult<()> {
+        for cert in certs {
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/transaction_manager.rs:68]: I think this is still necessary to avoid races.

**Review by mystenmark — APPROVED** [review]
Couple small comments left, happy to take the remaining issues on in subsequent PRs.

File: crates/sui-core/src/transaction_manager.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-core/src/transaction_manager.rs:68]: Added locking, and checks for if the transaction is already pending / executing / executed.

File: crates/sui-core/src/authority/authority_store.rs (hunk 18 above)
Comment (mwtian) [reply crates/sui-core/src/authority/authority_store.rs:367]: Can users upload a move package? Is it possible a node is just behind for the move package input? Added a comment but can abort the function with an error here instead.

File: crates/sui-core/src/authority/authority_store.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (mwtian) [reply crates/sui-core/src/authority/authority_store.rs:1242]: Added TODO.

File: crates/sui-core/src/authority.rs [hunk 21]
```diff
@@ -2317,18 +2365,18 @@ impl AuthorityState {
...
+                        .await?;
                 } else {
                     self.database
                         .record_owned_object_cert_from_consensus(&certificate, consensus_index)
-                        .await
+                        .await?;
                 }
+
+                let mut transaction_manager = self.transaction_manager.lock().await;
+                transaction_manager.enqueue(vec![certificate]).await
```
Comment (andll) [inline crates/sui-core/src/authority.rs:2379]: Question about this (apologies if this was answered somewhere else). I see that transaction_manager seem to be in-memory data structure. What happens if node restarts? Who guarantees that transaction will be executed eventually?

Comment (mwtian) [reply crates/sui-core/src/authority.rs:2379]: I mean they can be think of as the replay log for crash recovery of the TransactionManager, even though it is technically a map / rocksdb table. Basically we re-apply pending certificates on restart.

Comment (andll) [reply crates/sui-core/src/authority.rs:2379]: Sorry, I still don't fully follow. `TransactionManager::pending_certificates` is a BTreeMap, how can it be used to re-apply certificates on restart? Is the plan to change it to DBMap in the future?

Comment (andll) [reply crates/sui-core/src/authority.rs:2379]: Got it. And I see that we now write to `pending_certificates` in `finish_consensus_message_process`. Thanks for clarification!
